# COM400 Week 7

# Probability
Here we'll explore some ways we can use Python in the context of probability to carry out calculations that might be difficult or tedious to carry out by hand and to plot probability distributions. However, to start with we'll look at one or two simple examples of probability.

### Basic Probability
Example 6.3 from lecture notes.

Suppose $P(A) = 0.9, P(B) = 0.8$ and $P(A \cap B) = 0.75$. Find the probability that 

1. $A$ or $B$ occurs, i.e. $P(A \cup B)$
2. neither $A$ nor $B$ occurs, i.e. $P(A^c \cap B^c)$


In [5]:
P_A = 0.9
P_B = 0.8
P_AnB = 0.75 # represents the probability of the intersection of A and B

# Q1 - use addition rule
P_AuB = P_A + P_B - P_AnB # represents the probability of the union of A and B
print('Probability of A or B is {:5.3f}'.format(P_AuB))

# Q2 - use de Morgan's laws 
P_AcnBc = 1 - P_AuB
print('Probability of neither A nor B is {:5.3f}'.format(P_AcnBc))


Probability of A or B is 0.950
Probability of neither A nor B is 0.050


### Law of Total Probability

**You should read Section 6.7 of the lecture notes and ask in lab if you have any questions.**

Suppose $P(B|A) = 0.8$, $P(B|A^C) = 0.1$ and $P(A) = 0.2$. What is the probability of $B$?

Since $A$ and $A^c$ form a partition, we can use the law of total probability (see lecture notes) as follows

In [6]:
P_A = 0.2
P_BgA = 0.8 # represents the conditional probability of B given A
P_BgAc = 0.1

# Now use law of total probability
# Note that we need the probability of the complement of A which is 1 - P_A
P_B = P_BgA * P_A + P_BgAc * (1 - P_A)
print('Probability of B is {:5.3f}'.format(P_B))

Probability of B is 0.240


Suppose a company buys a particular component for its product from eight different
manufacturers denoted $M_1$, $M_2$, $\ldots$, $M_8$. The percentage it buys from each company is
30%, 22%, 15%, 12%, 8%, 6%, 4% and 3% respectively, while the percentage of components from each company that are defective is 1%, 2.5%, 0.3%, 1.3%, 0.9%, 1.7%, 0.6% and 1.2% respectively. If we randomly select one of the company’s product, what is the probability the component will be defective?

(See worked example 6.7 from the lecture notes for a similar problem.)

In [7]:
import numpy as np

# Use arrayes to represent the percentages for the manufacturers
# and for the probability of a defect given each manufacturer
P_M = np.array([[30, 22, 15, 12, 8, 6, 4, 3]])/100
P_FgM = np.array([[1, 2.5, 0.3, 1.3, 0.9, 1.7, 0.6, 1.2]])/100

# Now use law of total probability
# First multiply the probability of each manufacturer by the probability
# of a fault given that manufacturer
P_FnM = P_FgM * P_M
# Then add up all these probabilities to get the overall probability of a defect
P_F = np.sum(P_FnM)

print('Probability the component is defective: {:5.2f}%'.format(100 * P_F))

Probability the component is defective:  1.29%


### Bayes' Theorem

**You should read Section 6.8 of the lecture notes and ask in lab if you have any questions.**

(This example is similar to Example 6.8 from the notes.)

Suppose a medical test is positive for 94% of subjects who have a certain disease and negative for 96% of those who do not have the disease. Suppose also that 4% of the population have the disease. What is the probability that a randomly selected person who tested positive does not have the disease?

- Let $D$ be the event that the person has the disease. 
- Let $P$ be the event that the test result for the person is positive and $N$ that it is negative.

We can represent the information given to us as 

- $P(D) = 0.04$ and from the complement rule $P(D^c) = 0.96$
- $P(P|D) = 0.94$ and from the complement rule $P(N|D) = 0.06$
- $P(N|D^c) = 0.96$ and from the complement rule $P(P|D^c) = 0.04$

We want to find $P(D|P)$. We can use Bayes' theorem

\begin{equation}
P(D|P) = \frac{P(P|D)}{P(P)}P(D)
\end{equation}

However, we need use this formula we need $P(P)$. For it, we can use the law of total probability

\begin{equation}
P(P) = P(P|D)P(D) + P(P|D^c)P(D^c)
\end{equation}


In [8]:
P_D = 0.04
P_PgD = 0.94 # Probability of positive test result given the disease
P_NgDc = 0.96 # Probability of negative test result given no disease

# First we can get the probability of no disease using the complement rule
P_Dc = 1 - P_D

# We can also get the probability of a positive test result given no disease using the complement rule
P_PgDc = 1 - P_NgDc

# Now apply law of total probability to get the probability of a postive test result
P_P = P_PgD * P_D + P_PgDc * P_Dc

# Now apply Bayes' theorem to get the probability we want: P_DgP
P_DgP = (P_PgD * P_D) / P_P

print('Probability that the person has the disease, given the ' 
      'positive test result: {:5.3f}'.format(P_DgP))


Probability that the person has the disease, given the positive test result: 0.495


In examples like this we can think of Bayes' theorem as providing us with a way to update probabilities in light of new evidence. Here the **prior** (or initial) probability for the disease is $P(D) = 0.04$. In other words, this is the probability that a person has the disease before we take the test result into account. However, the positive test result provides evidence that the person does have the disease. When we apply Bayes' theorem, this gives us the **posterior** probability that the person has the disease (after taking the evidence into account) and this is $P(D|P) = 0.495$. 

Note, however, that it is still slightly more likely that the person does not have the disease! At this stage, further tests would be carried out. Suppose a second test is positive for 90% of subjects who have the disease and negative for 95% of those who do not have the disease. Suppose also that the person who underwent the first test and got a positive result, also gets a positive result for the second test. How could we apply Bayes' theorem?

The first thing to do is to take our **posterior** probability from the first test and make it the new **prior** probability, i.e. $P(D)$ is now given the value of $P(D|P)$, so $P(D) = 0.495$. [Strictly speaking this is a different probability distribution so we should use a different symbol like $P'(D) = 0.495$, but we'll not worry about that complication here.] Then we re-apply Bayes' theorem for the second test result. The code is almost identical.

In [9]:
# First set the posterior probability from the previous calculation to be the new prior probability
P_D = P_DgP 

# As before specify the conditional probabilities for the second test
P_PgD = 0.9 # Probability of positive test result given the disease
P_NgDc = 0.95 # Probability of negative test result given no disease

# First we can get the probability of no disease using the complement rule
P_Dc = 1 - P_D

# We can also get the probability of a positive test result given no disease using the complement rule
P_PgDc = 1 - P_NgDc

# Now apply law of total probability to get the probability of a postive test result
P_P = P_PgD * P_D + P_PgDc * P_Dc

# Now apply Bayes' theorem to get the probability we want: P_DgP
P_DgP = (P_PgD * P_D) / P_P

print('Probability that the person has the disease, given the ' 
      'second positive test result: {:5.3f}'.format(P_DgP))

Probability that the person has the disease, given the second positive test result: 0.946


Not surprisingly, we see that the probability that the person has the disease after the two positive test results is much higher. The process could then be continued to take account of further new evidence.

### Simulations based on Probability

The focus has been on determining the probability of certain outcomes, but probability is also used to simulate outcomes in science, engineering, finance, etc. Random numbers are used to do this. Suppose we want to simulate the outcome of tossing a fair coin. If we randomly select a number between 0 and 1 (with all numbers being equally likely to be chosen), then we would expect that around half the time (on half the trials) the number would be less than 0.5 and the rest of the time it would be greater than or equal to 0.5. So if the outcome is less than 0.5, we could use this to simulate a head (H) otherwise a tail (T). 


In [10]:
import numpy as np

p_head = 0.5

randNum = np.random.random()

if randNum < p_head:
    outcome = 'H'
else:
    outcome = 'T'

print('Coin flip outcome ' + outcome)

Coin flip outcome T


And this can easily be extended to multiple trials. So for ten trials:

In [11]:
for i in range(10):
    randNum = np.random.random()
    if randNum < p_head:
        outcome = 'H'
    else:
        outcome = 'T'
    print(outcome, end=' ')

T H H T H T T T T T 

A similar approach can be applied when the outcomes are are not equally likely. Suppose a coin is biased and has a probability of landing of heads of 0.3. We could simulate this be selecting a random number and letting it represent heads if is less than 0.3 and tails otherwise as follows:

In [12]:
p_head = 0.3

randNum = np.random.random()

if randNum < p_head:
    outcome = 'H'
else:
    outcome = 'T'

print('Coin flip outcome ' + outcome)

Coin flip outcome T


## Exercise 1
In an experiment $P(A) = \frac{1}{2}$, $P(B) = \frac{1}{2}$ and $P(A\cup B) = \frac{2}{3}$. Calculate:

i) $P(A^c)$

ii) $P(B^c)$

iii) $P(A\cap B)$

iv) $P(A^c \cap B^c)$

v) $P(A^c \cup B^c)$

vi) $P(A \cap B^c)$

vii) $P(A^c \cap B)$

viii) $P(A^c \cup B)$

In [13]:
# Stores the given values
P_A = 1/2
P_B = 1/2
P_AuB = 2/3

# Calculates the unknown values
P_Ac = 1 - P_A
P_Bc = 1 - P_B
P_AnB = P_A + P_B - P_AuB
P_AcnBc = 1 - P_AuB
P_AcuBc = 1 - P_AnB
P_AnBc = P_A - P_AnB
P_AcnB = P_B - P_AnB
P_AcuB = P_Ac + P_B - P_AcnB

# Outputs the results
print('i) P(Ac) = {:5.3f}'.format(P_Ac))
print('ii) P(Bc) = {:5.3f}'.format(P_Bc))
print('iii) P(A n B) = {:5.3f}'.format(P_AnB))
print('iv) P(Ac n Bc) = {:5.3f}'.format(P_AcnBc))
print('v) P(Ac u Bc) = {:5.3f}'.format(P_AcuBc))
print('vi) P(A n Bc) = {:5.3f}'.format(P_AnBc))
print('vii) P(Ac n B) = {:5.3f}'.format(P_AcnB))
print('viii) P(Ac u B) = {:5.3f}'.format(P_AcuB))

i) P(Ac) = 0.500
ii) P(Bc) = 0.500
iii) P(A n B) = 0.333
iv) P(Ac n Bc) = 0.333
v) P(Ac u Bc) = 0.667
vi) P(A n Bc) = 0.167
vii) P(Ac n B) = 0.167
viii) P(Ac u B) = 0.833


## Exercise 2
Suppose a bag contains 100 coins and that 40 coins are fair coins (i.e. the probability of heads is 1/2), 35 are biased so that their probability of heads is 0.4, and 25 are biased so that their probability of heads is 0.9. If a coin is selected at random and tossed what is the probability that it will be heads?

Hint: use the law of total probability.

In [14]:
import numpy as np

# Probabilities of picking each coin
P_C = np.array([[40,35,25]])/100
# Probabilities of heads for each coin
P_HgC = np.array([[0.5,0.4,0.9]])

# Multiplies the probability of each coin by the probability of heads given that coin
P_HnC = P_HgC * P_C
# Gets the overall probability of heads by summing the probabilities
P_H = np.sum(P_HnC)

# Outputs the results
print('Probability a toss of a random coin will be heads is {:5.2f}%'.format(100 * P_H))

Probability a toss of a random coin will be heads is 56.50%


## Exercise 3
Suppose there are two bags $A$ and $B$ each with 100 balls. $A$ has 50 white balls and 50 red balls, while $B$ has 90 white balls and 10 red balls. Both bags look identical from the outside and one is on the table in front of you. Although you aren't sure which it is, you have some reason to think that it is bag $A$ so you assign a prior probability of 0.7 to $A$. An experiment is set up as follows. You are allowed to take one ball at random from the bag (without looking inside) and observe whether it is white or red. When you do this, you observe a white ball. Use Bayes' theorem to determine the probability that the bag is $A$ and the probability that it is $B$ given this observation.

Hint: $P(A) = 0.7$ and $P(B) = 0.3$ are the prior probabilities. We also know the conditional probability of observing a white ball, which we can denote $W$, given that the bag is bag $A$. This probability is $P(W|A) = 0.5$. (Clearly, if $R$ represents observing a red ball, then $P(R|A) = 0.5$.) Similarly, we know that $P(W|B) = 0.9$. (Clearly, $P(R|B) = 0.1$.) We can now use the law of total probability to work out $P(W)$ and then Bayes' theorem to work out the probability that the bag is $A$ given that we have observed white, i.e. $P(A|W)$. Obviously, $P(B|W) = 1 - P(A|W)$.

In [15]:
# Stores given values
P_A = 0.7
P_B = 0.3
P_WgA = 0.5
P_WgB = 0.9

# Calculates the probabilty of getting white while having each bag
P_WnA = P_WgA * P_A
P_WnB = P_WgB * P_B
# Gets the overall probability by summing the probabilities
P_W = P_WnA + P_WnB

# Calculates the probability of each bag
P_AgW = (P_WgA * P_A)/P_W
P_BgW = 1 - P_AgW

# Outputs the results
print('The probability the bag is A is {:5.2f}%'.format(100*P_AgW))
print('The probability the bag is B is {:5.2f}%'.format(100*P_BgW))

The probability the bag is A is 56.45%
The probability the bag is B is 43.55%


## Exercise 4
After observing the white ball in exercise 3, the ball is then returned to the bag. Another ball is withdrawn randomly. It also is observed to be white and then returned to the bag. Then for a third and final time, a ball is withdrawn randomly and again it is observed to be white. Use Bayes' theorem to obtain the probability that the bag is $A$ and the probability that it is $B$ after all three observations.

Hint: The result after the first observation has already been obtained in exercise 3. We now need to take the posterior probabilities we obtained and set these to be the new prior probabilities. Then we start the process again to update the probabilities for $A$ and $B$ based on the second observation of a white ball. These then become our prior probabilities and we repeat the process once more.

In [16]:
# Runs twice
for i in range(2):
    # Stores known values
    P_A = P_AgW
    P_B = P_BgW

    # Uses law of total probability to get probability of white
    P_W = (P_A * P_WgA) + (P_B * P_WgB)

    # Uses Bayes' theorem to calculate probability of each bag
    P_AgW = (P_WgA * P_A)/P_W
    P_BgW = 1 - P_AgW

# Outputs the results
print('The probability the bag is A is {:5.2f}%'.format(100*P_AgW))
print('The probability the bag is B is {:5.2f}%'.format(100*P_BgW))

The probability the bag is A is 28.58%
The probability the bag is B is 71.42%


## Exercise 5
Consider one coin of each of three types in exercise 2, i.e. one coin is fair, another is biased so that its probability of heads is 0.4, and the third is biased so that its probability of heads is 0.9. Suppose you toss the fair coin. If it lands heads you then toss the second coin ten times, otherwise you toss the third coin ten times. Simulate this process.

Hint: You should first of all simulate the tossing of the first coin. You will need to select a random number to do this. Based on the outcome, your code should keep track of whether the second or third coin is to be used. Then you should toss the selected coin ten times (selecting a new random number each time), recording the outcome each time as H or T.

In [17]:
# Stores probabilities of heads given each coin
P_HgC = np.array([0.5,0.4,0.9])
# Creates string to store results
result = ''

# Tosses first coin and runs code depending
if np.random.random() < P_HgC[0]:
    # Adds toss to result
    result = 'H'
    # Runs 10 times
    for i in range(10):
        # Tosses coin and adds to result
        if np.random.random() < P_HgC[1]:
            result += ', H'
        else:
            result += ', T'
else:
    # Adds toss to result
    result = 'T'
    # Runs 10 times
    for i in range(10):
        # Tosses coin and adds to result
        if np.random.random() < P_HgC[2]:
            result += ', H'
        else:
            result += ', T'

# Outputs results
print(result)

H, H, H, T, T, T, T, T, H, H, H
